controller/tests/update_tests.ml

```diff
@@ -0,0 +1,99 @@
+open Update
+open Lwt
+open Update_test_helpers
+
+let happy_flow_test () =
+  let init_state = GettingVersionInfo in
+  let current_version = "10.0.1" in
+  let next_version = "10.0.2" in
+
+  let expected_bundle_name vsn =
+      Mock_update_client.test_bundle_name ^ _MAGIC_PAT ^ vsn ^ _MAGIC_PAT
+  in
+  let expected_url =
+    Config.System.update_url ^ _MAGIC_PAT
+        ^ expected_bundle_name next_version
+  in
+
+  let expected_state_sequence =
+    [
+      UpdateMock (fun () ->
+        Mock_rauc.set_status SystemA {
+            Mock_rauc.some_status with version = current_version
+        };
```

Comment:
You mean the test case should define the full `Rauc.Slot.status` record, rather than relying on some "base" record (`Rauc_mock.some_status`)? I see what you mean, but the only slot status field that is actually used by the `UpdateService` is the version, so everything else is just noise when it comes to the test specification.

Maybe `Mock_rauc` should just provide a method `set_version : slot -> version -> ()`, instead of having to deal with the whole status?

My general philosophy is that the test should be explicit _only_ about the things that are being tested in it. So if the test is not testing anything about the `status.device` or `status.installed_timestamp`, then why would it specify it? It makes reading/understanding the test harder, because you assume all specified values have some significance.

Reply:
> My general philosophy is that the test should be explicit _only_ about the things that are being tested in it. So if the test is not testing anything about the `status.device` or `status.installed_timestamp`, then why would it specify it? It makes reading/understanding the test harder, because you assume all specified values have some significance.

I see that point, but it also goes the other way: How do I know the test only relies on the explicitly specified data, and not other, shared properties? It is easy for a test author to accidentally rely on more than is made explicit, which in turn makes it harder to understand a test's intended assumptions.

Reply:
> How do I know the test only relies on the explicitly specified data, and not other, shared properties?

Well, you almost never know this, since the outcomes can (unintentionally) depend on arbitrary things (anything from developer-machine specific things, e.g. locale, to implementation quirks of the libraries in use). However, by being explicit you are at least saying "my assumption is that specified values / properties `A`, `B`, `C` fully determine the outcome". If this assumption is valid - all good, if not - well then you have a bug either in your implementation or your test spec, since you assumed some inputs are irrelevant, while they actually are.

However, I think you are maybe referring to problems with having many "shared" implicit assumptions in the (helper) test code, which is slightly a different issue.

I'll change it to use the full state, it's fine either way.

Reply:
Updated the test cases [to explicitly define the full RAUC slot status](https://github.com/dividat/playos/blob/aa8271017dd9a891404e4657ca31acd9706e6f3e/controller/tests/update_tests.ml#L14-L25) - is this how you see it done? It is [kinda repetitive](https://github.com/dividat/playos/blob/aa8271017dd9a891404e4657ca31acd9706e6f3e/controller/tests/update_tests.ml#L61-L67) and after adding N more test scenarios there will be an itch to generalize, at which point we would kinda go back square one.